rumqttc/src/state.rs

```diff
@@ -309,8 +322,13 @@ impl MqttState {
             .remove(&pubrec.pkid)
             .ok_or(StateError::Unsolicited(pubrec.pkid))?;
 
+        // Notify user about the publish, pubrel and pubcomp will be handled in background
+        if let Some(tx) = tx  {
+            tx.success();
+        }
```

Comment:
What is "handled in the background" supposed to mean in this context.

From what I understand, this makes the `NoticeFuture` complete on `PubRec` instead of `PubComp` for `ExactlyOnce` QoS, which is the wrong behavior from what I understand.

Reply:
IMO the notice is to confirm that the message actually received by broker.
So in QoS0 it's noticed on message sent (since there is no ACK), in QoS1 it's noticed on ACK reception, in QoS2 it's noticed on REC which indicates broker accept the message. The PubRel and PubComp do not affect the broker reception of message, but some way to let broker release QoS2 message.

Reply:
> QoS2 it's noticed on REC which indicates broker accept the message. The PubRel and PubComp do not affect the broker reception of message, but some way to let broker release QoS2 message.

The original code by @de-sh notifies on `PubComp` not on `PubRec` for QoS2 which I believe to be the correct behavior. Because when using this feature, I want to be notified once I'm certain the message will fulfil it's QoS. For QoS2, the message is only sent after sending `PubRel` to the broker, which the broker then acknowledges by responding with `PubComp`. If the client dies after receiving `PubRec` but before the broker received `PubRel`, the message might get lost and the `NoticeFuture` shouldn't resolve for a message that might not fulfil it's QoS, which for 1 and 2 means it not getting lost.

At least that's my take based on my understanding of how MQTT QoS works and my take on the `NoticeFuture` feature.

Reply:
What I mean is we can get notice on message ownership transfer, which is mentioned in the section 4.4, "When its original sender receives the PUBREC packet, ownership of the Application Message is transferred to the receiver". But I am totally fine to be noticed by PUBCOMP, where message ownership definitely transferred.

Reply:
BTW I see you working on migrating this ack wait feature to main in https://github.com/bytebeamio/rumqtt/pull/883, hoping the PR and discuss here can help.